syz-manager/manager.go

```diff
@@ -1066,7 +1061,143 @@ func (mgr *Manager) fuzzerConnect() ([]rpctype.RPCInput, BugFrames, bool) {
-	return corpus, BugFrames{memoryLeaks: memoryLeakFrames, dataRaces: dataRaceFrames}, mgr.coverFilterFilename != ""
+	var err error
+	mgr.once.Do(func() {
+		err = mgr.updateKernelModules(a)
+		if err != nil {
+			log.Fatalf("failed to update kernel modules: %v", err)
+			return
+		}
+		mgr.modulesInitialized = true
+		mgr.coverFilterBitmap, mgr.coverFilter, err = mgr.updateCoverFilter()
+	})
+	if err != nil {
+		return nil, BugFrames{}, nil, err
+	}
+	r.CoverFilterBitmap = mgr.coverFilterBitmap
+	return corpus, BugFrames{memoryLeaks: memoryLeakFrames, dataRaces: dataRaceFrames}, mgr.coverFilter, nil
+}
+
+const EXT string = ".ko"
+
+func searchModuleName(data []byte) (string, error) {
+	key := []byte("name=")
+	pos := bytes.Index(data, key)
+	if pos == -1 {
+		return "", fmt.Errorf("not find name= keyword")
+	}
+	end := bytes.Index(data[pos:], []byte{0x00})
+	if end == -1 {
+		return "", fmt.Errorf("not find null char")
+	}
+	return string(data[pos+len(key) : pos+end]), nil
+}
+
+func getModuleName(path string) (string, error) {
+	file, err := elf.Open(path)
+	if err != nil {
+		return "", fmt.Errorf("failed to open %v", path)
+	}
+	section := file.Section(".modinfo")
+	if section == nil {
+		return "", fmt.Errorf("no .modinfo section")
+	}
+	data, err := section.Data()
+	if err != nil {
+		return "", fmt.Errorf("failed to read .modinfo")
+	}
+	name, err := searchModuleName(data)
+	if err != nil {
+		section = file.Section(".gnu.linkonce.this_module")
+		if section == nil {
+			return "", fmt.Errorf("no .gnu.linkonce.this_module section")
+		}
+		data, err = section.Data()
+		if err != nil {
+			fmt.Println("failed to read .gnu.linkonce.this_module")
+		}
+		name = string(data)
+	}
+	return name, nil
+}
+
+func findModulePaths(dir string) ([]string, error) {
+	files := []string{}
+	err := filepath.Walk(dir, func(path string, f os.FileInfo, err error) error {
+		if filepath.Ext(path) == EXT {
+			files = append(files, path)
+		}
+		return nil
+	})
+
+	return files, err
+}
+
+func autoFindModulePath(where, name string) (string, error) {
+	paths, err := findModulePaths(where)
+	if err != nil {
+		return "", err
+	}
+	for _, path := range paths {
+		if strings.TrimSuffix(filepath.Base(path), EXT) == name {
+			return path, nil
+		}
+	}
+	for _, slowpath := range paths {
+		mname, err := getModuleName(slowpath)
+		if err != nil {
+			return "", err
+		}
+		if mname == name {
+			return slowpath, nil
+		}
+	}
+	return "", fmt.Errorf("no match module path found")
+}
+
+func (mgr *Manager) getModulePath(name string) (string, error) {
+	modulePath := ""
+	var err error
+	if mgr.cfg.KernelObj != "" {
+		modulePath, err = autoFindModulePath(mgr.cfg.KernelObj, name)
+	}
+	for _, path := range mgr.cfg.ModuleObj {
+		modulePath, err = autoFindModulePath(path, name)
+		if err == nil {
+			return modulePath, nil
+		}
+	}
+	if err != nil {
+		return "", err
+	}
+	return modulePath, err
+}
+
+func (mgr *Manager) updateKernelModules(a *rpctype.ConnectArgs) error {
```

Comment:
All of the modules code need to be moved to some package because:
1. The main manager file is a bad place to throw in all assorted logic we have.
2. This logic is Linux-specific. syz-manager code is not OS-specific. It is common for 9 different OSes.
3. This does not make it possible to wrap it into a tool. For this code it may be useful to have a tool that scans a given dir and prints all discovered modules and their names (testing this by running whole syz-manager is very painful).

Of all packages we currently have pkg/cover itself looks like the best fit for this. So perhaps we just pass in list of unresolved modules and module dirs to cover.MakeReportGenerator and do this logic in pkg/cover.